sQTL analysis with Leafcutter using genotype data from previous Lieber eQTL experiments
========

### Step 1: Locate & list all junction .count files
Run `01_input_check.R` to get filenames of the count files from rse_gene, check 
that all exist, then export to `data/junc_count_files.txt` 

### Step 2: convert .count files to .junc files
Starting from the junction count files generated by featureCounts, we create .junc files 
that have a different column order but are otherwise identical

Run with `02_reformat_junc.sh`
```
for i in $(ls /dcl01/lieber/ajaffe/lab/goesHyde_mdd_rnaseq/preprocessed_data/Counts/junction/*.count); 
	do python ../scripts/reformat_junc.py $i;
done
```

This creates a file containing the names of all .junc files
`ls ../data/jucn/*.junc >> ../data/all_jxn_filenames.txt`

### Step 3: cluster introns
Generate intron clusters from the .junc files, this results in many different files,
but the one used in step 3 is a matrix of all junction/individual pairs where
each cell is a ratio in string format "#/#"

Run with `03_leafcutter_cluster.sh`
```
python leafcutter_cluster.py -j ../data/all_jxn_filenames.txt -m 50 -l 500000 -r ../data/clusters/
```

### Step 4: Generate a normalized ratio matrix
Generates normalized ration matrix files for each chromosome (qqnorm files) and non normalized files
Also generates a file containing covariates controlled by the -p flag (eg -p 10 = 10 covariates)
We will generate our own covariates in line with the MatrixEQTL experiments in future steps

Run with `04_generate_ratio_matrix.sh`
Output: `data/clusters/leafcutter_perind.counts.gz.qqnorm_chr*.gz`

### Step 5: Convert qqnorm files to bed.gz files for each region
Load all `data/clusters/leafcutter_perind.counts.gz.qqnorm_chr*.gz` files, combine across chromosomes, 
then seperate by brain region, format for tensorqtl. Combined data has 210544 clusters.

Run with `05_extract_qqnorm_data.R`
Output: `data/qqnorm/qqnorm_Amygdala.bed.gz` & `data/qqnorm/qqnomr_sACC.bed.gz`

### Step 6: Generate covariate files
Use code from the MatrixeQTL analysis to regenerate a genotype vcf and run
PCA on the qqnorm ratio matrix

Run with: `06_calculate_covariates.R`

### Step 7: tensorQTL
Run `cis.nominal` for all splice QTL results 
genotype: All SNPs

phenotype: `data/qqnorm/qqnorm_*.bed.gz`
covariates: qqnorm PCA + model
